AK/SipHash.cpp

```diff
@@ -0,0 +1,168 @@
+    u64 v2 = 0x6c7967656e657261ull;
+    u64 v3 = 0x7465646279746573ull;
+    auto const left = input.size() & 7;
+    // The end of 64-bit blocks.
+    auto const block_end = input.size() - (input.size() % sizeof(u64));
+    u64 b = input.size() << 56;
+    v3 ^= key.high();
+    v2 ^= key.low();
+    v1 ^= key.high();
+    v0 ^= key.low();
+
+    if (output.size() == 16)
+        v1 ^= 0xee;
+
+    FixedMemoryStream input_stream { input };
+    for (size_t input_index = 0; input_index < block_end; input_index += 8) {
+        // The way we use them, no stream functions can fail.
+        auto const m = MUST(input_stream.read_value<LittleEndian<u64>>());
+        v3 ^= m;
+
+        for (size_t i = 0; i < message_block_rounds; ++i)
+            SIPROUND;
+
+        v0 ^= m;
+    }
+
+    switch (left) {
+    case 7:
+        b |= (static_cast<u64>(input[block_end + 6])) << 48;
+        [[fallthrough]];
+    case 6:
+        b |= (static_cast<u64>(input[block_end + 5])) << 40;
+        [[fallthrough]];
+    case 5:
+        b |= (static_cast<u64>(input[block_end + 4])) << 32;
+        [[fallthrough]];
+    case 4:
+        b |= (static_cast<u64>(input[block_end + 3])) << 24;
+        [[fallthrough]];
+    case 3:
+        b |= (static_cast<u64>(input[block_end + 2])) << 16;
+        [[fallthrough]];
+    case 2:
+        b |= (static_cast<u64>(input[block_end + 1])) << 8;
+        [[fallthrough]];
+    case 1:
+        b |= (static_cast<u64>(input[block_end + 0]));
+        break;
+    case 0:
+        break;
+    }
+
+    v3 ^= b;
+
+    for (size_t i = 0; i < message_block_rounds; ++i)
+        SIPROUND;
+
+    v0 ^= b;
+
+    if (output.size() == 16)
+        v2 ^= 0xee;
+    else
+        v2 ^= 0xff;
+
+    for (size_t i = 0; i < finalization_rounds; ++i)
+        SIPROUND;
+
+    b = v0 ^ v1 ^ v2 ^ v3;
+
+    FixedMemoryStream output_stream { output };
+    MUST(output_stream.write_value<LittleEndian<u64>>(b));
+
+    if (output.size() == 8)
+        return;
+
+    v1 ^= 0xdd;
+
+    for (size_t i = 0; i < finalization_rounds; ++i)
+        SIPROUND;
+
+    b = v0 ^ v1 ^ v2 ^ v3;
+    MUST(output_stream.write_value<LittleEndian<u64>>(b));
+}
+
+// Using a shared key greatly simplifies DynamicLoader and Kernel,
+// since they don't need to worry about synchronizing access to block-scope statics.
+#ifdef KERNEL
+static u128 sip_hash_key = Kernel::get_good_random<u128>();
+#else
+// get_random is assumed to be secure, otherwise SipHash doesn't deliver on its promises!
+static u128 sip_hash_key = get_random<u128>();
+#endif
+
+template<size_t message_block_rounds, size_t finalization_rounds>
+unsigned SipHash<message_block_rounds, finalization_rounds>::operator()(u64 input)
+{
+    ReadonlyBytes input_bytes { &input, sizeof(input) };
+    auto key = sip_hash_key;
+    u64 output;
+    Bytes output_bytes { &output, sizeof(output) };
+    do_siphash<message_block_rounds, finalization_rounds>(input_bytes, key, output_bytes);
+    return output;
```

Comment:
This implicitly chops off some bits, that's no good!

Reply:
Yup, forgot to XOR the two halves.